src/xml_parser_main_metadata_together.c

```diff
@@ -404,9 +416,7 @@ cr_PkgIterator_parse_next(cr_PkgIterator *iter, GError **err) {
     cr_CbData *cbdata = (cr_CbData*) iter->cbdata;
 
     while (!cr_PkgIterator_is_finished(iter) && g_queue_is_empty(cbdata->finished_pkgs_queue)) {
-        while ((cbdata->in_progress_count_primary <= cbdata->in_progress_count_filelists ||
-                cbdata->in_progress_count_primary <= cbdata->in_progress_count_other) &&
-               !iter->primary_is_done)
+        while (((cbdata->in_progress_count_primary <= cbdata->in_progress_count_filelists) && (cbdata->loadingflags & CR_PACKAGE_LOADED_FIL) || (cbdata->in_progress_count_primary <= cbdata->in_progress_count_other) && (cbdata->loadingflags & CR_PACKAGE_LOADED_OTH) || cbdata->loadingflags == CR_PACKAGE_LOADED_PRI) && !iter->primary_is_done)
```

Comment:
Ew.  Suggestions welcome on how to clean this up.

Reply:
I can take another look at it, but I had trouble getting the conditionals such that it wouldn't infinite loop or terminate too early.

Reply:
The tests seem to pass but I worry that this isn't going to work if the file doesn't fit completely inside the first XML buffer.  The conditional is "&&" so when it comes back around it's never going to move forwards if the filelists or other in progress haven't moved from 0.

Reply:
Well, I did a test with a smaller buffer and it still seems to work so I guess I'm missing something.

Reply:
Some problems with infinite looping or terminating too early were fixed by the `|` -> `&` switch up.

> The tests seem to pass but I worry that this isn't going to work if the file doesn't fit completely inside the first XML buffer. The conditional is `&&`, so when it comes back around after the first buffer gets processed it's never going to move forwards because `in_progress_primary_count > 0` and `in_progress_{filelists|other} == 0`.

No, the `in_progress_primary_count` doesn't count packages that have not finished parsing primary. It counts packages that have fully parsed primary but are still in progress (require filelists/other to be parsed). The same goes for `in_progress_filelists_count` and `in_progress_other_count`.
I now see that the naming is confusing. :disappointed:

Reply:
> Though both the proposed version and the unchanged version will parse the whole file at once if just `primary` is provided (loading it all to memory at once). 🤔

I was thinking how to make this work and one solution I came up with is to change:
```diff
-        while ((cbdata->in_progress_count_primary <= cbdata->in_progress_count_filelists ||
-                cbdata->in_progress_count_primary <= cbdata->in_progress_count_other) &&
-                !iter->primary_is_done)
-        {
+        if (!iter->primary_is_done) {
```

This makes sense to me because `primary` is mandatory so the main loop parses `primary` and the nested `while` loops just make sure `filelists` and `other` catch up if needed.

To further simplify it I think we could also remove the comparisons of `in_progress_count_other` and `in_progress_count_filelists` like:
```diff
             }
         }

-        while ((cbdata->in_progress_count_filelists <= cbdata->in_progress_count_primary ||
-                cbdata->in_progress_count_filelists <= cbdata->in_progress_count_other) &&
-               !iter->filelists_is_done)
-        {
+        while ((cbdata->in_progress_count_filelists <= cbdata->in_progress_count_primary) && !iter->filelists_is_done) {
             iter->filelists_is_done = parse_next_section(iter->filelists_f, iter->filelists_path, iter->filelists_pd, err);
             if (*err) {
                 return NULL;
             }
         }

-        while ((cbdata->in_progress_count_other <= cbdata->in_progress_count_filelists ||
-                cbdata->in_progress_count_other <= cbdata->in_progress_count_primary) &&
-               !iter->other_is_done)
-        {
+        while ((cbdata->in_progress_count_other <= cbdata->in_progress_count_primary) && !iter->other_is_done) {
             iter->other_is_done = parse_next_section(iter->other_f, iter->other_path, iter->other_pd, err);
             if (*err) {
                 return NULL;
```

This changes ordering of warnings in a couple of tests but those can be amended.

It also makes this whole `while` madness a bit simpler.

Reply:
Seems like it does more than change the ordering?

Reply:
Ok, now there is no reordering, it seems the buffer size change (which was in your temp commit) is also needed for it to happen.